mono/metadata/marshal.c

```diff
@@ -758,11 +758,33 @@ mono_string_builder_new (int starting_string_length, MonoError *error)
 static void
 mono_string_utf16_to_builder_copy (MonoStringBuilderHandle sb, const gunichar2 *text, size_t string_len, MonoError *error)
 {
-	MonoArrayHandle chunkChars = MONO_HANDLE_NEW_GET (MonoArray, sb, chunkChars);
-	gchandle_t gchandle = 0;
-	memcpy (MONO_ARRAY_HANDLE_PIN (chunkChars, gunichar2, 0, &gchandle), text, sizeof (gunichar2) * string_len);
-	mono_gchandle_free_internal (gchandle);
-	MONO_HANDLE_SETVAL (sb, chunkLength, int, string_len);
+	MonoArrayHandle chunkChars = MONO_HANDLE_NEW (MonoArray, NULL);
+	MonoStringBuilderHandle chunk = MONO_HANDLE_NEW (MonoStringBuilder, MONO_HANDLE_RAW (sb));
+
+	guint capacity = mono_string_builder_capacity (sb);
+
+	g_assert (capacity >= string_len);
+
+	MONO_ENTER_NO_SAFEPOINTS;
+
+	do {
+		MONO_HANDLE_GET (chunkChars, chunk, chunkChars);
+		const int maxLength = MONO_HANDLE_GETVAL (chunkChars, max_length);
+		g_assert (maxLength >= 0);
+		const int chunkOffset = MONO_HANDLE_GETVAL (chunk, chunkOffset);
+		g_assert (chunkOffset >= 0);
+		if (maxLength > 0 && chunkOffset < string_len) {
+			// Check that we will not overrun our boundaries.
+			int charsToCopy = MIN (string_len - chunkOffset, maxLength);
+			memcpy (MONO_HANDLE_RAW (chunkChars)->vector, text + chunkOffset, charsToCopy * sizeof (gunichar2));
+			MONO_HANDLE_SETVAL (chunk, chunkLength, int, charsToCopy);
+		} else {
+			MONO_HANDLE_SETVAL (chunk, chunkLength, int, 0);
+		}
+		MONO_HANDLE_GET (chunk, chunk, chunkPrevious);
+	} while (MONO_HANDLE_BOOL (chunk));
+
+	MONO_EXIT_NO_SAFEPOINTS;
```

Comment:
This code is quite long and admittedly not easy to read. Alternative is to go in CoreCLR footsteps and rewrite the function to always spit out `StringBuilder` with single chunk.